src/dfx/src/commands/deps/pull.rs

```diff
@@ -0,0 +1,342 @@
+            );
+            pulled_json
+                .canisters
+                .insert(canister_id, PulledCanister::default());
+        }
+    }
+    Ok(())
+}
+
+// download canister wasm, candid, init
+async fn download_canister_files(
+    env: &dyn Environment,
+    logger: &Logger,
+    canister_id: Principal,
+    pulled_canister: &mut PulledCanister,
+) -> DfxResult {
+    info!(logger, "Pulling canister {canister_id}...");
+
+    let agent = env
+        .get_agent()
+        .ok_or_else(|| anyhow!("Cannot get HTTP client from environment."))?;
+
+    // try fetch `dfx:wasm_hash`. If not available, get the hash of the on chain canister.
+    let hash_on_chain = match fetch_metadata(agent, canister_id, DFX_WASM_HASH).await? {
+        Some(wasm_hash_raw) => {
+            let wasm_hash_str = String::from_utf8(wasm_hash_raw)?;
+            trace!(
+                logger,
+                "Canister {canister_id} specified a custom hash: {wasm_hash_str}"
+            );
+            hex::decode(wasm_hash_str)?
+        }
+        None => {
+            match read_state_tree_canister_module_hash(agent, canister_id).await? {
+                Some(hash_on_chain) => hash_on_chain,
+                None => {
+                    bail!("Canister {canister_id} doesn't have module hash. Perhaps it's not installed.");
+                }
+            }
+        }
+    };
+
+    pulled_canister.wasm_hash = hex::encode(&hash_on_chain);
+
+    // will save wasm and candid in $(cache_root)/pulled/{canister_id}/
+    let canister_dir = get_cache_root()?
+        .join("pulled")
+        .join(canister_id.to_string());
+    dfx_core::fs::create_dir_all(&canister_dir)?;
+
+    let wasm_path = get_pulled_wasm_path(canister_id)?;
+
+    // skip download if cache hit
+    let mut cache_hit = false;
+    let wasm_url_txt_path = get_pulled_wasm_url_txt_path(canister_id)?;
+    if wasm_path.exists() {
+        let bytes = dfx_core::fs::read(&wasm_path)?;
+        let hash_cache = Sha256::digest(bytes);
+        if hash_cache.as_slice() == hash_on_chain {
+            cache_hit = true;
+            trace!(logger, "The canister wasm was found in the cache.");
+            let wasm_url_str = dfx_core::fs::read_to_string(&wasm_url_txt_path)
+                .with_context(|| format!("Failed to read {:?}", &wasm_url_txt_path))?;
+            pulled_canister.wasm_url = Some(wasm_url_str);
+        }
+    }
+    if !cache_hit {
+        // fetch `dfx:wasm_url`
+        let wasm_url_raw = fetch_metadata(agent, canister_id, DFX_WASM_URL)
+            .await?
+            .ok_or_else(|| {
+                anyhow!("`{DFX_WASM_URL}` metadata not found in canister {canister_id}.")
+            })?;
+        let wasm_url_str = String::from_utf8(wasm_url_raw)?;
+        let wasm_url = reqwest::Url::parse(&wasm_url_str)?;
+        write_to_tempfile_then_rename(wasm_url_str.as_bytes(), &wasm_url_txt_path)?;
```

Comment:
This is the only place that uses `dfx:wasm_url`, and it uses it right away, so do we really need to store it in pulled.json?